modules/xphoto/src/grayworld_white_balance.cpp

```diff
@@ -0,0 +1,232 @@
+
+            v_max1 = v_cvt_f32(v_reinterpret_as_s32(v_max(v_iB1, v_max(v_iG1, v_iR1))));
+            v_max2 = v_cvt_f32(v_reinterpret_as_s32(v_max(v_iB2, v_max(v_iG2, v_iR2))));
+            v_max3 = v_cvt_f32(v_reinterpret_as_s32(v_max(v_iB3, v_max(v_iG3, v_iR3))));
+            v_max4 = v_cvt_f32(v_reinterpret_as_s32(v_max(v_iB4, v_max(v_iG4, v_iR4))));
+
+            v_sat1 = (v_max1 - v_min1) / v_max1;
+            v_sat2 = (v_max2 - v_min2) / v_max2;
+            v_sat3 = (v_max3 - v_min3) / v_max3;
+            v_sat4 = (v_max4 - v_min4) / v_max4;
+
+            // Calculate masks
+            v_m1 = v_reinterpret_as_u32(v_sat1 <= v_thresh);
+            v_m2 = v_reinterpret_as_u32(v_sat2 <= v_thresh);
+            v_m3 = v_reinterpret_as_u32(v_sat3 <= v_thresh);
+            v_m4 = v_reinterpret_as_u32(v_sat4 <= v_thresh);
+
+            // Apply mask
+            v_SB += (v_iB1 & v_m1) + (v_iB2 & v_m2) + (v_iB3 & v_m3) + (v_iB4 & v_m4);
+            v_SG += (v_iG1 & v_m1) + (v_iG2 & v_m2) + (v_iG3 & v_m3) + (v_iG4 & v_m4);
+            v_SR += (v_iR1 & v_m1) + (v_iR2 & v_m2) + (v_iR3 & v_m3) + (v_iR4 & v_m4);
+        }
+
+        // Perform final reduction
+        sum1 = v_reduce_sum(v_SB);
+        sum2 = v_reduce_sum(v_SG);
+        sum3 = v_reduce_sum(v_SR);
+#endif
+        double minRGB, maxRGB, satur;
+        for ( ; i < N3; i += 3 )
+        {
+            minRGB = min(src_data[i], min(src_data[i + 1], src_data[i + 2]));
+            maxRGB = max(src_data[i], max(src_data[i + 1], src_data[i + 2]));
+            satur = (maxRGB - minRGB) / maxRGB;
+            if ( satur > thresh ) continue;
+            sum1 += src_data[i];
+            sum2 += src_data[i + 1];
+            sum3 += src_data[i + 2];
+        }
+
+        // Find inverse of averages
+        double dinv1 = sum1 == 0 ? 0.f : (double)N / (double)sum1,
+               dinv2 = sum2 == 0 ? 0.f : (double)N / (double)sum2,
+               dinv3 = sum3 == 0 ? 0.f : (double)N / (double)sum3;
+
+        // Find maximum
+        double inv_max = max(dinv1, max(dinv2, dinv3));
+
+        // Convert to floats
+        float inv1 = (float) dinv1,
+              inv2 = (float) dinv2,
+              inv3 = (float) dinv3;
+
+        // Scale by maximum
+        if ( inv_max > 0 )
+        {
+            inv1 = (float)((double)inv1 / inv_max);
+            inv2 = (float)((double)inv2 / inv_max);
+            inv3 = (float)((double)inv3 / inv_max);
+        }
+
+        // Scale input pixel values
+        uchar* dst_data = dst.ptr<uchar>(0);
+        i = 0;
+#if CV_SIMD128
+        v_uint8x16  v_in, v_out;
+        v_uint32x4  v_i1, v_i2, v_i3, v_i4;
+        v_float32x4 v_f1, v_f2, v_f3, v_f4,
+                    scal1(inv1, inv2, inv3, inv1),
+                    scal2(inv2, inv3, inv1, inv2),
+                    scal3(inv3, inv1, inv2, inv3),
+                    scal4(inv1, inv2, inv3, 0.f);
+
+        for ( ; i < N3 - 14; i += 15 )
```

Comment:
inv1, inv2, inv3 are all within [0, 1] range. could you please scale the channels in fixed-point arithmetics instead of going floating-point?